### PR TITLE
Tests | Introduce RAII SQL object primitives

### PR DESCRIPTION
**Pull request by edwardneal**
## Description

This change is designed to make sure that the manual tests clean up after themselves and to introduce some consistency.

Many of the manual tests need to create objects - usually tables, stored procedures and UDTs. There are a handful of ways to do this. Some of the AlwaysEncrypted tests have a set of objects describing tables, column encryption keys, etc. The `DataTestUtility` class has `CreateSP`, `DropSP` and similar methods. Some tests have their own helper methods and others manually construct DDL statements. These are also used in various combinations, and most of the time the logic to delete these temporary objects is located in a `finally` block.

Separately, the objects are created in different scopes - sometimes they're passed as fixtures to the test class, sometimes they're created by the constructor of a test class which implements `IDisposable` but most of the time they're created by each test.

Lastly, the naming of these is sometimes odd. Sometimes the tests request a random object name, other times they'll use existing helper methods in `DataTestUtility`, and other times still they'll attempt to use their own logic to generate a unique name.

I'd like to simplify this with a consistent set of RAII-style types which represent database objects, returning its name post-hoc.

This PR starts the process by introducing types for a table, a UDT or an SP. It then applies these to a handful of tests (ApiShould, JsonStreamTest, JsonTest, SqlGraphTable, ParametersTest) to prove that they handle the edge cases. In a few cases (the two changed JsonStream tests) we eliminate test bugs which would always leave a JSON-based table behind, or which would leave files behind on the filesystem if an exception was thrown at the wrong time.

I expect it to take a while to migrate to the RAII types, but it should cut down on resource leaks in the Azure databases.

## Issues

None.

## Testing

Automated tests should continue to pass.

### CODE REVIEW

**Comment by ErikEJ** [comment]
What does RAII mean exactly?

**Comment by edwardneal** [comment]
Resource Acquisition Is Initialization - so we create SQL objects when a `Table` / `StoredProcedure` / `UserDefinedType` class is instantiated, and drop the SQL object when the instance is disposed of.

**Review by cheenamalhotra — APPROVED** [review]
LGTM

File: src/Microsoft.Data.SqlClient/tests/Common/Fixtures/DatabaseObjects/DatabaseObject.cs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        const int MaxWaits = 2;
+        int counter = MaxWaits;
+
+        if (Connection.State is System.Data.ConnectionState.Closed)
+        {
+            Connection.Open();
+        }
+        while (counter-- > 0 && Connection.State is System.Data.ConnectionState.Connecting)
+        {
+            Thread.Sleep(80);
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/Common/Fixtures/DatabaseObjects/DatabaseObject.cs:49]: Nit: Explain this magic number?

Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/tests/Common/Fixtures/DatabaseObjects/DatabaseObject.cs:49]: I kept this the same as [`DataTestUtility.ResurrectConnection`](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs#L862-L872), which was added as part of #1658 without explanation. I assume that the method was added to make sure that CI remains stable when the target SQL Server instance is under sufficient load to drop/time out connections, but I'm not sure why it's sleeping for 80ms specifically (perhaps trial and error?)

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/ParametersTest.cs [hunk 2]
```diff
@@ -262,67 +247,41 @@ public static void Test_WithGuidValue_ShouldReturnGuid()
...
-                    dbData.Load(reader);
-                    Assert.Equal(1, dbData.Rows.Count);
-                    Assert.Equal(x, dbData.Rows[0][0]);
-                    Assert.Equal(y, dbData.Rows[0][1]);
-                }
-            }
-            finally
-            {
-                using SqlCommand cmd = connection.CreateCommand();
-                cmd.CommandText = "DROP PROCEDURE " + procName;
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/ParametersTest.cs:319]: This makes me think, won't we need an ordering guarantee to make sure that all tables that use a UDT are dropped before the type is dropped (and same for other object dependencies)? Otherwise we may dispose out of order and fail to clean something up.

**Review by mdaigle — CHANGES_REQUESTED** [review]

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/ParametersTest.cs (hunk 2 above)
Comment (edwardneal) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/ParametersTest.cs:319]: If we have table A, user-defined table type B and stored procedure C, we'll only be able to run their creation DML in the order `A / B / C` or `B / A / C` - the stored procedure won't compile until the objects it references are present. At that point, we've got to create corresponding primitives A, then B, then C. As long as we let C# handle the disposal via `using` blocks, they'll be disposed in the reverse order, and this handles the simple case.

There are a few places where we create the objects in the test class constructor or inject it using fixtures. These objects aren't always SQL Server primitives. In the Always Encrypted tests, we've got one fixture which creates a certificate, then a column encryption key, then a series of tables (and I think, a stored procedure.)

I think we should keep the object primitives' responsibility simple, and provide a disposable base class for fixtures like this which pushes the underlying objects onto a stack as they're created. Upon disposal, the base class can pop the objects off the stack and dispose of them. The existing [`SQLSetupStrategy`](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/TestFixtures/SQLSetupStrategy.cs) fixture does something similar to this (although it definitely needs rework - it's fairly specialised to AE, it's injected into a lot of tests, each of those tests only uses a small number of the tables, and it takes on responsibility for running itself against multiple connection strings.)

If you're happy with the approach, I'd prefer to do that in a future PR, once the simple cases (a test creating its own tables/etc.) are dealt with and there's a workable set of tests remaining to prove the design.

Comment (mdaigle) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/ParametersTest.cs:319]: I wan't aware that there was a guaranteed ordering to disposal, so this all makes sense with that context. I'm happy with the approach.

**Review by mdaigle — APPROVED** [review]
